### DISCUSSION TITLE
debugging advice for main.lua errors

### DISCUSSION THREAD

**Discussion by anielsen001** [quarto-dev/quarto-cli#9334] [Q&A]
### Description

I'm getting an error when I try to render to a jupyter notebook, but i don't get the error if I just make a webpage. It looks like some kind of formatting issue in my qmd file. Is there a way to find out more details about where the error is? I see line numbers that refer to the `main.lua` script, but nothing referring to my inputs. Is there a way to find out where the error is so that I can try to address it? 

The error looks like this:
```
[INFO] Running filter /opt/quarto/share/filters/main.lua
Error running filter /opt/quarto/share/filters/main.lua:
Block, list of Blocks, or compatible element expected, got table
	while retrieving function argument content
	while retrieving arguments for function Div
stack traceback:
	/opt/quarto/share/filters/main.lua:20822: in field 'render'
	/opt/quarto/share/filters/main.lua:835: in local 'filter_fn'
	/opt/quarto/share/filters/main.lua:302: in function </opt/quarto/share/filters/main.lua:292>
	(...tail calls...)
	[C]: in ?
	[C]: in method 'walk'
	/opt/quarto/share/filters/main.lua:224: in function </opt/quarto/share/filters/main.lua:214>
	(...tail calls...)
	/opt/quarto/share/filters/main.lua:936: in local 'callback'
	/opt/quarto/share/filters/main.lua:954: in upvalue 'run_emulated_filter_chain'
	/opt/quarto/share/filters/main.lua:990: in function </opt/quarto/share/filters/main.lua:987>
stack traceback:
	/opt/quarto/share/filters/main.lua:224: in function </opt/quarto/share/filters/main.lua:214>
	(...tail calls...)
	/opt/quarto/share/filters/main.lua:936: in local 'callback'
	/opt/quarto/share/filters/main.lua:954: in upvalue 'run_emulated_filter_chain'
	/opt/quarto/share/filters/main.lua:990: in function </opt/quarto/share/filters/main.lua:987>
```

**Comment by mcanouil** [comment]
Could you share a small **self-contained "working" (reproducible)** example to work with, _i.e._, a complete Quarto document or a Git repository? Thanks.

The usual debugging step is to make a reproducible example because it makes it easier to see what input is causing this (it could be a Quarto bug).

Note that sharing an error code without context is unlikely to get you with a concrete solution (here and anywhere else).

See https://quarto.org/docs/troubleshooting/index.html for the troubleshooting guide.



---

You can share a **self-contained "working" (reproducible)** Quarto document using the following syntax, _i.e._, using more backticks than you have in your document (usually four ` ```` `).
See <https://quarto.org/bug-reports.html#small-is-beautiful-aim-for-a-single-document-with-10-lines>.

If you have multiple files (and if it is **absolutely required** to have multiple files), please share as a Git repository.

<table>
<tr><th>R</th><th>Python</th></tr>
<tr><td>

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
engine: knitr
---

This is a reproducible Quarto document.

```{r}
x <- c(1, 2, 3, 4, 5)
y <- c(1, 4, 9, 16, 25)

plot(x, y)
```

![An image](https://placehold.co/600x400.png)

The end.
````
`````

</td><td>

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
engine: jupyter
---

This is a reproducible Quarto document.

```{python}
import matplotlib.pyplot as plt

x = [1, 2, 3, 4, 5]
y = [1, 4, 9, 16, 25]

plt.plot(x, y)
plt.show()
```

![An image](https://placehold.co/600x400.png)

The end.
````
`````

</td></tr>
</table>

Additionally and if not already given, please share the output of `quarto check` within a code blocks (*i.e.*, using three backticks ` ```txt `), see <https://quarto.org/bug-reports.html#check>.

**Comment by anielsen001** [comment]
I will work to make one. The file I'm working with is quite large and complex. I was hoping to find a way to narrow down how to locate the problem within my file to create an example. 

As a bit of context, I've been working on this as a html output for a while. The computations are complex and when I also output a ipynb format, quarto recomputes everything twice, so I disable the ipynb output until I'm done, then I reenable the ipynb output for the final product. because of this I don't know where to start looking for the issue. I can go thru and remove each section one at a time, but that will take a while.

I guess my question is if there is a way to find the location of the error in my code.

**Reply by mcanouil** [reply]
From the error message, I would assume it's related to a table in the wrong place which make me think are you using Quarto 1.3?

**Reply by anielsen001** [reply]
No, here's my quarto check output.

```
quarto check
Quarto 1.4.549
[✓] Checking versions of quarto binary dependencies...
      Pandoc version 3.1.11: OK
      Dart Sass version 1.69.5: OK
      Deno version 1.37.2: OK
[✓] Checking versions of quarto dependencies......OK
[✓] Checking Quarto installation......OK
      Version: 1.4.549
      Path: /opt/quarto/bin

[✓] Checking tools....................OK
      TinyTeX: (not installed)
      Chromium: (not installed)

[✓] Checking LaTeX....................OK
      Using: Installation From Path
      Path: /usr/bin
      Version: 2023

[✓] Checking basic markdown render....OK

[✓] Checking Python 3 installation....OK
      Version: 3.11.8
      Path: /opt/project/venv_sp2024/bin/python3
      Jupyter: 5.7.2
      Kernels: python3

[✓] Checking Jupyter engine render....OK

[✓] Checking R installation...........OK
      Version: 4.3.2
      Path: /usr/lib/R
      LibPaths:
        - /usr/local/lib/R/site-library
        - /usr/lib/R/site-library
        - /usr/lib/R/library
      knitr: 1.45
      rmarkdown: 2.25

[✓] Checking Knitr engine render......OK

```

**Reply by mcanouil** [reply]
So, Unlikely to be what I had in mind.

Small trick for you if you are using knitr.
Add a code cell with: `knitr::knit_exit()`
This will make the rendering happens until this command, so you could move this toward the end of your document (or backward).
Workflow:
1. put it in the middle
2. if it fails: move it to the middle of the first half of your document
   if it succeed: move it to the middle of the second half of your document
3. iterate using this approach to get down to the part leading to the issue (which I still think is a syntax issue with a table and/or a Quarto bug regarding that table)

**Comment by cderv** [comment]
To help debug, you could also activate Verbose mode and stack trace. It could say more that the current error
https://quarto.org/docs/troubleshooting/#verbose-mode

You could also try latest stable version 1.4, and possibly dev version of Quarto in case this is fixed already

**Comment by cscheid** [comment]
(And, just to be clear, this is a Quarto bug! no crashes like that should happen.)

**Reply by cscheid** [reply]
I've tracked down the crash to https://github.com/quarto-dev/quarto-cli/blob/main/src/resources/filters/customnodes/floatreftarget.lua#L935

Somehow, the `float.content` field is getting corrupted.

It has to do with one of your crossreferenceable floats ("Figure", "Table", etc), so you could start by removing those selectively.

**Reply by mbellitti** [reply]
I'm having the same issue on Quarto 1.5.54, removing all the `#| label: ` tags from my document makes the rendering go through. How can I help debug this further? I would like to use cross referencing for figures.

**Reply by cscheid** [reply]
@mbellitti A small reproducible example is the best way you can help.

**Reply by Steinthal** [reply]
I got a similar error yesterday after pasting in some text from Quicklook and could figure out what the cause was (in my case): The quotation marks in »widths="1.6"« (Quarto 1.5.55). Here's the the reproducible example:

````imd
---
title: "main.lua error"
format: pdf
---

# Test

::: {.list-table widths=“1,6“ header-rows=0}
* - n/a
  - Not applicable, not relevant

:::
````

**Reply by mcanouil** [reply]
You need straight double quotes.
Other symbols are not valid (that's not a Quarto specificity).

**Reply by Steinthal** [reply]
I know. That’s why I wrote the quotation mark was the cause of that error. It’s just meant as one example that produces the error or at least a similar error.

**Reply by mcanouil** [reply]
FYI, VSCode and other IDEs can detect these issues likely coming from copy/paste.

**Reply by Steinthal** [reply]
Thanks, that’s good to know. I’ll check Positron for that feature.

**Comment by mbellitti** [comment]
The following example throws an error on quarto 1.5.54

<pre>
---
title: "Minimal example"
format: gfm
---

```{python}
#| label: fig-sin
import matplotlib.pyplot as plt
import numpy as np

x = np.linspace(0,1,10)
y = np.sin(x)

plt.plot(x,y)
```
</pre>

The error is the following:

```
Error running filter /home/opt/quarto-1.5.54/share/filters/main.lua:
Block, list of Blocks, or compatible element expected, got table
        while retrieving function argument blocks
        while retrieving arguments for function Blocks
stack traceback:
        /home/opt/quarto-1.5.54/share/filters/main.lua:22460: in field 'render'
        /home/opt/quarto-1.5.54/share/filters/main.lua:1233: in local 'filter_fn'
        /home/opt/quarto-1.5.54/share/filters/main.lua:635: in function </home/opt/quarto-1.5.54/share/filters/main.lua:625>
        (...tail calls...)
        [C]: in ?
        [C]: in method 'walk'
        /home/opt/quarto-1.5.54/share/filters/main.lua:557: in function </home/opt/quarto-1.5.54/share/filters/main.lua:546>
        (...tail calls...)
        /home/opt/quarto-1.5.54/share/filters/main.lua:1334: in local 'callback'
        /home/opt/quarto-1.5.54/share/filters/main.lua:1352: in upvalue 'run_emulated_filter_chain'
        /home/opt/quarto-1.5.54/share/filters/main.lua:1388: in function </home/opt/quarto-1.5.54/share/filters/main.lua:1385>
stack traceback:
        /home/opt/quarto-1.5.54/share/filters/main.lua:557: in function </home/opt/quarto-1.5.54/share/filters/main.lua:546>
        (...tail calls...)
        /home/opt/quarto-1.5.54/share/filters/main.lua:1334: in local 'callback'
        /home/opt/quarto-1.5.54/share/filters/main.lua:1352: in upvalue 'run_emulated_filter_chain'
        /home/opt/quarto-1.5.54/share/filters/main.lua:1388: in function </home/opt/quarto-1.5.54/share/filters/main.lua:1385>
```

It sounds like it has something to do with the `gfm` output format, it works correctly if I render to html. The first time I encountered this bug I was rendering my project to both gfm and html, and changing that to just render html works as expected. I was using gfm because it gives a nice preview on GitHub.

**Reply by cderv** [reply]
Thanks for the report. I can reproduce and I opened an issue for this.

**Reply by cderv** [reply]
- https://github.com/quarto-dev/quarto-cli/issues/10358
